src/ws.cpp

```diff
@@ -532,6 +532,7 @@ void Instance::unregisterViewBackend(uint32_t surfaceId)
     auto it = m_viewBackendMap.find(surfaceId);
     if (it != m_viewBackendMap.end()) {
         it->second->apiClient = nullptr;
+        wl_client_destroy(it->second->client);
```

Comment:
I'm seeing a crash in this call that leads me to believe that this client has already been destroyed elsewhere when this is called. Here is the stacktrace:

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  wl_list_insert_list (list=list@entry=0x6f00b528, other=other@entry=0x6f00b520) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/wayland-util.c:89
89            other->next->prev = list;
[Current thread is 1 (LWP 16)]
(gdb) back
#0  wl_list_insert_list (list=list@entry=0x6f00b528, other=other@entry=0x6f00b520) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/wayland-util.c:89
#1  0x703b8ac4 in wl_priv_signal_emit (signal=signal@entry=0x6f00b520, data=data@entry=0x6f00b4e0) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/wayland-server.c:1992
#2  0x703b8c70 in wl_client_destroy (client=0x6f00b4e0) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/wayland-server.c:852
#3  0x6fb140d4 in WS::Instance::unregisterViewBackend (this=0x6f01d530, surfaceId=1) at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/ws.cpp:673
#4  0x6fb123a4 in ViewBackend::unregisterSurface (this=this@entry=0x6f0b37c0, surfaceId=<optimized out>)
    at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/view-backend-exportable-private.cpp:137
#5  0x6fb124bc in ViewBackend::unregisterSurface (surfaceId=<optimized out>, this=0x6f0b37c0) at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/view-backend-exportable-private.cpp:49
#6  ViewBackend::~ViewBackend (this=0x6f0b37c0, __in_chrg=<optimized out>) at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/view-backend-exportable-private.cpp:45
#7  0x6fb15e74 in <lambda(void*)>::operator() (__closure=<optimized out>, data=<optimized out>) at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/view-backend-exportable-private.h:123
#8  <lambda(void*)>::_FUN(void *) () at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/view-backend-exportable-private.h:124
#9  0x73ec3f38 in wpe_view_backend_destroy (backend=0x6f0b1bd8) at /usr/src/debug/libwpe/1.6.0-r0/libwpe-1.6.0/src/view-backend.c:59
#10 0x6fb11e60 in wpe_view_backend_exportable_fdo_destroy (exportable=0x6f0b18e8) at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/view-backend-exportable-fdo.cpp:182
#11 0x74714518 in void WTF::derefGPtr<_WebKitWebViewBackend>(_WebKitWebViewBackend*) () from /opt/veld/1.0.0/sysroots/cortexa9hf-neon-veld-linux-gnueabi/usr/lib/libWPEWebKit-1.0.so.3.10.6
#12 0x7470d6e6 in _WebKitWebViewPrivate::~_WebKitWebViewPrivate() () from /opt/veld/1.0.0/sysroots/cortexa9hf-neon-veld-linux-gnueabi/usr/lib/libWPEWebKit-1.0.so.3.10.6
#13 0x7470bddc in webkit_web_view_finalize(_GObject*) () from /opt/veld/1.0.0/sysroots/cortexa9hf-neon-veld-linux-gnueabi/usr/lib/libWPEWebKit-1.0.so.3.10.6
#14 0x74009df4 in g_object_unref (_object=0x6f0b5080) at ../glib-2.58.3/gobject/gobject.c:3346
#15 0x763e8240 in cog_shell_remove_view (shell=shell@entry=0x6f007de0, name=<optimized out>) at /usr/src/debug/cog/git-r0/git/core/cog-shell.c:526
```
The client itself in frame 2 looks like an already freed chunk of memory:

```
(gdb) print *client
$23 = {connection = 0x9, source = 0xa, display = 0x25, display_resource = 0x0, id_count = 0, mask = 36, link = {prev = 0x0, next = 0x0}, objects = {client_entries = {size = 18, alloc = 50, data = 0x0}, 
    server_entries = {size = 0, alloc = 0, data = 0x0}, side = 0, free_list = 0}, destroy_signal = {listener_list = {prev = 0x0, next = 0x0}, emit_list = {prev = 0x0, next = 0x0}}, ucred = {pid = 0, uid = 0, 
    gid = 0}, error = 33, resource_created_signal = {listener_list = {prev = 0x6f000044, next = 0x6f0740a8}, emit_list = {prev = 0x6f00b548, next = 0x6f00b548}}}
```

Reply:
That line of code was added apparently to fix some crash that it was happening at some point in the "Test code" attached in 
https://bugs.webkit.org/show_bug.cgi?id=218612. I can not reproduce anymore this crash even revert this commit so probably it is a superfluous change actually not needed.

Reply:
Should we revert the changes from this PR, then?

Reply:
Reverting 9a0e6cb62bfae44ef3496d950899ecb34ddd24a5 I found this new crash:

```
#1  0x703c0b04 in wl_priv_signal_emit (signal=signal@entry=0x6f0f6960, data=data@entry=0x6f0f6920) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/wayland-server.c:2008
#2  0x703c0c70 in wl_client_destroy (client=client@entry=0x6f0f6920) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/wayland-server.c:852
#3  0x703c0d4c in wl_client_connection_data (fd=<optimized out>, mask=<optimized out>, data=0x6f0f6920) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/wayland-server.c:341
#4  0x703c2a7c in wl_event_loop_dispatch (loop=0x6f00f818, timeout=timeout@entry=0) at /usr/src/debug/wayland/1.13.0-r0/wayland-1.13.0/src/event-loop.c:423
>>>>   #5  0x6fb1a758 in WS::ServerSource::<lambda(GSource*, GSourceFunc, gpointer)>::operator() (__closure=0x0, base=0x6f028780) at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/ws.cpp:120
#6  WS::ServerSource::<lambda(GSource*, GSourceFunc, gpointer)>::_FUN(GSource *, GSourceFunc, gpointer) () at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/ws.cpp:129
#7  0x73f2f668 in g_main_dispatch (context=0x1402740) at ../glib-2.58.3/glib/gmain.c:3182
#8  g_main_context_dispatch (context=context@entry=0x1402740) at ../glib-2.58.3/glib/gmain.c:3874
#9  0x73f2f88c in g_main_context_iterate (context=0x1402740, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib-2.58.3/glib/gmain.c:3947
#10 0x73f2fc2c in g_main_loop_run (loop=0x13e42c0) at ../glib-2.58.3/glib/gmain.c:4143
#11 0x004aea70 in wpe::WpeThreadedLauncher::runThread (this=this@entry=0x1404de8) at /usr/src/debug/annonymous-wpe-launcher/999-r0/git/src/WpeThreadedLauncher.cc:1005
#12 0x004ba45c in wpe::WpeThreadedLauncher::<lambda(void*)>::operator() (__closure=0x0, data=<optimized out>) at /usr/src/debug/annonymous-wpe-launcher/999-r0/git/src/WpeThreadedLauncher.cc:585
#13 wpe::WpeThreadedLauncher::<lambda(void*)>::_FUN(void *) () at /usr/src/debug/annonymous-wpe-launcher/999-r0/git/src/WpeThreadedLauncher.cc:589
#14 0x73f5bbdc in g_thread_proxy (data=0x140d000) at ../glib-2.58.3/glib/gthread.c:784
#15 0x76624e8c in start_thread (arg=0x6f9ff2b0) at /usr/src/debug/glibc/2.26-r0/git/nptl/pthread_create.c:465
#16 0x73cc7448 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from /mnt/data-external/image/sysroots/cortexa9hf-neon-image-linux-gnueabi/lib/libc-2.26.so
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```



Both crashes are happening in the `wl_client_destroy()`. The one that you found is in the `WS::Instance::unregisterViewBackend`,  the new one it is happening in the `wl_event_loop_dispatch()`. Both in the `wayland-server.c`  

AFAIK the UI-process is acting as Wayland server for the WebProcess so what is happending in both cases is, the wl_client_destroy() invoked for a wl_client already closed. I think, I saw it in the code in the past, the wl_client is closed in by wayland if there is a error in the communication (ex: the FD is closed because the other peer of the waylannd channel is gone, aka: WebProcess dead). Probably what we are missing a safe-guard (like `if (m_client.object)`) in some place.
    

The worklow looks like the following:


1. The WebProcess terminates and (as ws-client) sends an UnregisterSurface IPC message:

    ```
   usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/ws-client.cpp:        m_glib.socket->send(FdoIPC::Messages::UnregisterSurface, m_wl.wpeBridgeId);
    ```

2. The UI-process handles the message:

``` cpp
void ViewBackend::didReceiveMessage(uint32_t messageId, uint32_t messageBody)
{
    switch (messageId) {
    case FdoIPC::Messages::RegisterSurface:
        registerSurface(messageBody);
        break;
    case FdoIPC::Messages::UnregisterSurface:
        unregisterSurface(messageBody); <<<<<<<<<<
        break;
    default:
        assert(!"WPE fdo received an invalid IPC message");
    }
}
```

3. This also means: `ViewBackend::unregisterSurface()` -> `Instance::unregisterViewBackend()`


   (maybe the wl_client should be in the `ViewBackend::unregisterSurface()` and not in the `Instance`, because it is this first one who knows the existence of this wl_client).

3. a) (**with** 9a0e6cb62bfae44ef3496d950899ecb34ddd24a5) The `wl_client_destroy(it->second->client)` the idea was to prevent this client been called in the next loop iteration of the WS GSource  (`m_source`):
    ```
   #7  0x6fb15e74 in <lambda(void*)>::operator() (__closure=<optimized out>, data=<optimized out>) at /usr/src/debug/wpebackend-fdo/1.6.1-r0/wpebackend-fdo-1.6.1/src/view-backend-exportable-private.h:123
   ```

   ``` cpp
    // dispatch
    [](GSource* base, GSourceFunc, gpointer) -> gboolean                                                               
    {                                                                                                                  
        auto& source = *reinterpret_cast<ServerSource*>(base);                                                         
                                                                                                                       
        if (source.pfd.revents & G_IO_IN) {
            struct wl_event_loop* eventLoop = wl_display_get_event_loop(source.display);                               
            wl_event_loop_dispatch(eventLoop, 0); <<<<<<<<<<<<<<<<<<
            wl_display_flush_clients(source.display);                                                                  
        }
                                                                                                                       
        if (source.pfd.revents & (G_IO_ERR | G_IO_HUP))                                                                
            return FALSE;

        source.pfd.revents = 0;
        return TRUE;
    },
    ```

    but in your case this `wl_client_destroy()` call comes too late leading in a double destroy action.

3. b) (**without** 9a0e6cb62bfae44ef3496d950899ecb34ddd24a5), the GSource loop is dispatching events to a wl_client what it was already freed but still being referenciated for someone else (the queued GSource event?) so we are accessing to a memory what it could be free or not and in any case not valid anymore.

The solution could be something like safe-guard the `wl_event_loop_dispatch()` with something like `if (m_client.object)` but the dispatch events for all the backends registered in the WS::Instance so ...